kmip/services/server/engine.py

```diff
@@ -0,0 +1,404 @@
+        else:
+            raise exceptions.InvalidMessage(
+                "KMIP {0} is not supported by the server.".format(
+                    protocol_version
+                )
+            )
+
+    def _verify_credential(self, request_credential, connection_credential):
+        # TODO (peterhamilton) Add authentication support
+        # 1. If present, verify user ID of connection_credential is valid user.
+        # 2. If present, verify request_credential is valid credential.
+        # 3. If both present, verify that they are compliant with each other.
+        # 4. If neither present, set server to only allow Query operations.
+        pass
+
+    @_synchronize
+    def process_request(self, request, credential=None):
+        """
+        Process a KMIP request message.
+
+        This routine is the main driver of the KmipEngine. It breaks apart and
+        processes the request header, handles any message errors that may
+        result, and then passes the set of request batch items on for
+        processing. This routine is thread-safe, allowing multiple client
+        connections to use the same KmipEngine.
+
+        Args:
+            request (RequestMessage): The request message containing the batch
+                items to be processed.
+            credential (Credential): A credential containing any identifying
+                information about the client obtained from the client
+                certificate. Optional, defaults to None.
+
+        Returns:
+            ResponseMessage: The response containing all of the results from
+                the request batch items.
+        """
+        header = request.request_header
+
+        # Process the protocol version
+        self._set_protocol_version(header.protocol_version)
+
+        # Process the maximum response size
+        max_response_size = None
+        if header.maximum_response_size:
+            max_response_size = header.maximum_response_size.value
+
+        # Process the time stamp
+        if header.time_stamp:
+            then = header.time_stamp.value
+            now = int(time.time())
+
+            if (now >= then) and ((now - then) < 60):
+                self._logger.info("Received request at time: {0}".format(
+                    time.strftime(
+                        "%Y-%m-%d %H:%M:%S",
+                        time.gmtime(then)
+                    )
+                ))
+            else:
+                self._logger.warning(
+                    "Received request with old timestamp. Possible replay "
+                    "attack. {0}, {1}".format(then, now)
+                )
+                raise exceptions.InvalidMessage(
+                    "Stale request rejected by server."
+                )
+        else:
+            self._logger.info("Received request at time: {0}".format(
+                time.strftime(
+                    "%Y-%m-%d %H:%M:%S",
+                    time.gmtime(int(time.time()))
```

Comment:
I should. I added the time stamp now/then error checking after I added this log line.